### PR TITLE
[ENH] Confidence Interval improvements

### PR DESCRIPTION
**Pull request by Tillsten**
The CI calculation now uses the root_scalar interface for root finding
and changes the default method to toms748. Additonally, caching of
probabilties is now used to speed up the calculation.

#### Description

Improves the CI calculation by using some caching and using another root finding method.

###### Type of Changes

- [ ] Bug fix
- [ ] New feature
- [x] Refactoring / maintenance
- [ ] Documentation / examples

###### Tested on

Python: 3.9.12 (main, Apr  4 2022, 05:22:27) [MSC v.1916 64 bit (AMD64)]

lmfit: latest master, scipy: 1.7.1, numpy: 1.20.3, asteval: 0.9.23, uncertainties: 3.1.6

###### Verification 
Have you

- [ ] included docstrings that follow PEP 257?

- [ ] referenced existing Issue and/or provided relevant link to mailing list?

- [x] verified that existing tests pass locally?

- [x] verified that the documentation builds locally?

- [x] squashed/minimized your commits and written descriptive commit messages?

- [x] added or updated existing tests to cover the changes?
- [ ] updated the documentation and/or added an entry to the release notes (doc/whatsnew.rst)?
- [ ] added an example?

### CODE REVIEW

**Comment by reneeotten** [comment]
Sorry, busy days... I'll take a look tonight!

File: lmfit/confidence.py [hunk 1]
```diff
@@ -249,17 +255,12 @@ def calc_ci(self, para, direction):
...
-                                 limit, rtol=.5e-4, args=prob)
-                except ValueError:
-                    val = np.nan
-
+            sol = root_scalar(calc_prob, method='toms748', bracket=sorted([limit, a_limit]), rtol=.5e-4, args=(prob,))
+            if sol.converged:
+                val = sol.root
+            else:
+                val = np.nan
+                break
```
Comment (reneeotten) [inline lmfit/confidence.py:263]: sorry, it's late so I might be missing something.... why is the `break` needed here? There isn't a loop or anything going on so it seems redundant to me

**Review by reneeotten — COMMENTED** [review]
one comment on the code. Other than that the change is probably fine, I haven't investigated the advantage of `root_scalar` over `brentq` for this purpose, but I assume there is....?

Before merging (or when merging) please make sure to squash the commits; there is now a variety of random commits including a merg-commit in the history which shouldn't be there. Additionally, adding an entry to `whatsnew.rst` would probably be good, highlighting why this change is important and what the advantage is.

**Comment by newville** [comment]
@reneeotten Thanks!  And yes, I believe the idea that `root_scaler` is both the "more recent, supported" root-finding method, and that `toms748` is almost always faster (# of function calls) than the `brent` method. 

And I think that `break` at the test for "has solution converged" is within a for loop, expressing "if a solution is not converged don't bother trying to bracket the uncertainties for the other probabilities".

File: lmfit/confidence.py (hunk 1 above)
Comment (reneeotten) [reply lmfit/confidence.py:263]: sorry, as Matt pointed out it is actually in a "for-loop" - so my comment is wrong...

**Comment by newville** [comment]
@reneeotten any objections to merging this?

**Comment by newville** [comment]
@Tillsten OK, merging this, thanks!

**Comment by reneeotten** [comment]
> @reneeotten any objections to merging this?

not really, other than the comments that I left earlier: to squash the commits to remove the flurry of commits including a "merge-commit" and adding an entry to `whatsnew.rst` indicating this change.

**Pull request merged**